File: test/javascript/principles/transposer.test.js
import { scaleOf, chordsOf, transpose } from 'principles/transposer';

test('scaleOf', () => {
  expect(scaleOf('C')).toEqual(['C', 'D', 'E', 'F', 'G', 'A', 'B']);
  expect(scaleOf('A')).toEqual(['A', 'B', 'C#', 'D', 'E', 'F#', 'G#']);
  expect(scaleOf('Cm')).toEqual(['C', 'D', 'Eb', 'F', 'G', 'Ab', 'Bb']);
  expect(scaleOf('F#m')).toEqual(['F#', 'G#', 'A', 'B', 'C#', 'D', 'E']);
});

test('transpose', () => {
  expect(transpose({ from: 'C', to: 'G', note: 'C' })).toEqual('G');
  expect(transpose({ from: 'C', to: 'G', note: 'B' })).toEqual('F#');
  expect(transpose({ from: 'F', to: 'F#', note: 'D' })).toEqual('D#');
  expect(transpose({ from: 'F', to: 'Gb', note: 'D' })).toEqual('Eb');
  expect(transpose({ from: 'Bb', to: 'D', note: 'A#' })).toEqual('D');
});

test('chordsOf', () => {
  expect(chordsOf('C')).toEqual(['C', 'Dm', 'Em', 'F', 'G', 'Am']);
  expect(chordsOf('B')).toEqual(['B', 'C#m', 'D#m', 'E', 'F#', 'G#m']);
  expect(chordsOf('Fm')).toEqual(['Fm', 'Ab', 'Bbm', 'Cm', 'Db', 'Eb']);
  expect(chordsOf('Em')).toEqual(['Em', 'G', 'Am', 'Bm', 'C', 'D']);
});

File: test/lib/transposer_test.rb
require 'test_helper'
require 'transposer'

class TransposerTest < ActiveSupport::TestCase
  include Transposer

  test 'scale_of' do
    assert_equal %w[C D E F G A B], scale_of('C')
    assert_equal %w[A B C# D E F# G#], scale_of('A')
    assert_equal %w[C D Eb F G Ab Bb], scale_of('Cm')
    assert_equal %w[F# G# A B C# D E], scale_of('F#m')
  end

  test 'transpose' do
    assert_equal 'F#', transpose(from: 'C', to: 'G', note: 'B')
    assert_equal 'D#', transpose(from: 'F', to: 'F#', note: 'D')
    assert_equal 'Eb', transpose(from: 'F', to: 'Gb', note: 'D')
    assert_equal 'D', transpose(from: 'Bb', to: 'D', note: 'A#')
  end

  test 'chords_of' do
    assert_equal %w[C Dm Em F G Am], chords_of('C')
    assert_equal %w[B C#m D#m E F# G#m], chords_of('B')
    assert_equal %w[Fm Ab Bbm Cm Db Eb], chords_of('Fm')
    assert_equal %w[Em G Am Bm C D], chords_of('Em')
  end
end
